### PR TITLE
23 test ng

### LINKED ISSUE #23: [13] TestNG (with Flink) Introduction for the Team
### As a
Developer

### I want
to know how tests with testng have to be written.

### Because
I need some guidelines and a general overview about testNG and how it interacts with flink/Kafka including some examples in varying complexity

## Tasks
- [x] Pull Request open 
- [x] Add some example Tests as best practices in different complexity
- [x] ..including mocking events for flink
- [x] suggest how and what should be tested in relation to flink, we don't want to test flink itself but our code, but how exactly?
- [x] tests for accessing the DB for Planned data, find out if necessary
- [x] generate ideas what might be needed to be mocked to which extend in relation to our pipeline
- [x] Suggest some "test style standards"
- [x] Suggest a structure for Test components 
- [x] gain knowledge about different best practices
- [x] generate a short presentation share you insights with the Team

## Acceptance Criteria
- [ ] Documented
- [ ] on Master
- [ ] Linter is okay
- [ ] Tests are written and green
- [ ] Code Review done
- [ ] Reviewed by PO

### PR DESCRIPTION
**Pull request by ma-ro**
closes #23 by adding some test cases aswell as the DataProviders

### CODE REVIEW

File: src/main/java/org/bptlab/cepta/StreamingJob.java [hunk 1]
```diff
@@ -57,8 +68,61 @@ public static void main(String[] args) throws Exception {
...
     // execute program
-    env.execute("Flink Streaming Java API Skeleton");
+    env.execute("Start Streaming");
+  }
+
+  public static DataStream<Integer> stormFilter(DataStream<Integer> windspeed) {
+    return windspeed.filter(
+        new FilterFunction<Integer>() {
+          @Override
+          public boolean filter(Integer speed) throws Exception {
```
Comment (skydivin4ng3l) [inline src/main/java/org/bptlab/cepta/StreamingJob.java:79]: my java is not on expert level but i would expect a try with an exception and not directly an exception in the normal case? Is this the normal case?

File: src/main/java/org/bptlab/cepta/StreamingJob.java [hunk 2]
```diff
@@ -57,8 +68,61 @@ public static void main(String[] args) throws Exception {
...
+                  public boolean filter(Integer last, Context<Integer> context) throws Exception {
+                    for (Integer middle : context.getEventsForPattern("middle")) {
+                      if (middle > last) {
+                        return false;
+                      }
+                    }
+                    return true;
+                  }
+                });
+    PatternStream<Integer> patternStream = CEP.pattern(seaLevels, pattern);
```
Comment (skydivin4ng3l) [inline src/main/java/org/bptlab/cepta/StreamingJob.java]: are both pattern "variables" the Pattern<Integer, ?> pattern from line 88? if not naming should be improved. it feels a bit odd though.

File: pom.xml [hunk 3]
```diff
@@ -66,6 +66,12 @@ under the License.
 			<artifactId>flink-streaming-java_${scala.binary.version}</artifactId>
 			<version>${flink.version}</version>
 		</dependency>
+		<!-- Apache Flink Cep Engine -->
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-cep_2.11</artifactId>
+			<version>1.9.0</version>
```
Comment (skydivin4ng3l) [inline pom.xml:73]: should this also be a symbol like ${...} is this to be the flink version?

File: src/test/java/org/bptlab/cepta/DataProviderMultipleRisingDataStream.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.bptlab.cepta;
+
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.testng.annotations.DataProvider;
+
+public class DataProviderMultipleRisingDataStream {
```
Comment (skydivin4ng3l) [inline src/test/java/org/bptlab/cepta/DataProviderMultipleRisingDataStream.java:7]: Rising in which context? Rising in terms of increasing?

File: src/test/java/org/bptlab/cepta/DataProviderSingleRisingDataStream.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package org.bptlab.cepta;
+
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.testng.annotations.DataProvider;
+
+public class DataProviderSingleRisingDataStream {
```
Comment (skydivin4ng3l) [inline src/test/java/org/bptlab/cepta/DataProviderSingleRisingDataStream.java:7]: Single in which context? I see 3 DataStreams all in the dataProviderMethod, each with multiple elements? This is confusing, or not well explained in the documentation.

File: src/test/java/org/bptlab/cepta/DataProviderMultipleRisingDataStream.java (hunk 4 above)
Comment (skydivin4ng3l) [reply src/test/java/org/bptlab/cepta/DataProviderMultipleRisingDataStream.java:7]: Like rising SeaLevel?

Comment (skydivin4ng3l) [reply src/test/java/org/bptlab/cepta/DataProviderMultipleRisingDataStream.java:7]: Like Raising the alarm?

File: src/test/testng.xml [hunk 6]
```diff
@@ -6,6 +6,10 @@
         <groups>
             <define name="include-group">
                 <include name="active" />
+                <include name="include-test-one" />
+                <include name="include-test-two" />
```
Comment (skydivin4ng3l) [inline src/test/testng.xml]: where are these to defined? Are these supposed to be the pattern tests? or why are these here?

**Review by skydivin4ng3l — CHANGES_REQUESTED** [review]
Please also provided a bit more explaination concerning the Event nature of Data Streams, for me at least it is not 100% clear how the data streams fromCollections are interpreted.   are the intereger int he DataStreams send from left to right and how are the Object  [][] {{windspeed1}, {windspeed2}, {windspeed3}}; interpreted. All the events from windspeed1 then all the events from windspeed2,...etc? Those points remain a bit unclear.

File: src/main/java/org/bptlab/cepta/StreamingJob.java (hunk 1 above)
Comment (ma-ro) [reply src/main/java/org/bptlab/cepta/StreamingJob.java:79]: Exception in this function is unnecessary anyways - going to be removed.

File: src/main/java/org/bptlab/cepta/StreamingJob.java (hunk 2 above)
Comment (ma-ro) [reply src/main/java/org/bptlab/cepta/StreamingJob.java]: Don't quite know what two variables you mean. 

CEP. pattern is a function call and (seaLevels, pattern) the parameters. And yes the "pattern" in the parameter is the variable from line 88.

File: src/test/java/org/bptlab/cepta/DataProviderMultipleRisingDataStream.java (hunk 4 above)
Comment (ma-ro) [reply src/test/java/org/bptlab/cepta/DataProviderMultipleRisingDataStream.java:7]: Fair - this test is for the rising seaLevel.

File: src/test/java/org/bptlab/cepta/DataProviderSingleRisingDataStream.java (hunk 5 above)
Comment (ma-ro) [reply src/test/java/org/bptlab/cepta/DataProviderSingleRisingDataStream.java:7]: A rising sequenze in the context of the seaLevel method is defined as 3 directly following increasing integers.

In the SingleRising DataProvider each Stream has a single rising sequenze of events while the other DataProvider has streams with multiple rising sequenzes.

File: src/test/testng.xml (hunk 6 above)
Comment (ma-ro) [reply src/test/testng.xml]: that was from the original example - is removed

File: src/main/java/org/bptlab/cepta/StreamingJob.java (hunk 2 above)
Comment (skydivin4ng3l) [reply src/main/java/org/bptlab/cepta/StreamingJob.java]: hope you see the not optimal naming here

File: pom.xml (hunk 3 above)
Comment (ma-ro) [reply pom.xml:73]: travis has some problems when we use ${flink-version} so i would suggest we remain 1.9.0 and clearify this later

**Review by skydivin4ng3l — APPROVED** [review]
